src/util/key.rs

```diff
@@ -109,6 +109,21 @@ impl PublicKey {
         }
     }
 
+    /// Read the public key from a reader
+    pub fn read_from<R: io::Read>(mut reader: R) -> Result<Self, io::Error> {
+        let mut bytes = [0; 65];
+        let byte_sl;
+        reader.read_exact(&mut bytes[0..1])?;
+        if bytes[0] < 4 {
+            reader.read_exact(&mut bytes[1..33])?;
+            byte_sl = &bytes[0..33];
+        } else {
+            reader.read_exact(&mut bytes[1..65])?;
+            byte_sl = &bytes[0..65];
+        }
```

Comment:
I believe that `let bytes = if ...;` expression is considered idiomatic.

Or perhaps more concisely:

```rust
let mut bytes = [0; 65];
// using ..1 instead of 0..1 avoids needless bound check, so I tend to prefer it
reader.read_exact(&mut bytes[..1])?;
// shadowing is considered fine here
let bytes = if bytes[0] < 4 {
    &mut bytes[..33]
} else {
    &mut bytes[..65]
}
};

reader.read_exact(&mut bytes[1..])?;
Self::from_slice(bytes).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
```

BTW since this reads a single byte at the beginning, perhaps add a doc comment recommending to use buffered reader.

Reply:
+1 for the doc comment.

Reply:
If the compiler generates different code for `0..1` than `..1` then I'll eat my hat :)

Everything else I agree with.

Reply:
pushed; took your code pretty-much verbatim and added a doccomment

Reply:
> If the compiler generates different code for 0..1 than ..1 then I'll eat my hat :)

Yeah, I believe it should be same in release mode. It's just a random quirk in my brain to prefer it. :)